test/es-module/test-esm-dynamic-import-assertion.js

```diff
@@ -0,0 +1,68 @@
+// Flags: --experimental-json-modules
+'use strict';
+const common = require('../common');
+const { rejects, strictEqual } = require('assert');
+
+const jsModuleDataUrl = 'data:text/javascript,export{}';
+
+async function test() {
+  await rejects(
+    import(`data:text/javascript,import${JSON.stringify(jsModuleDataUrl)}assert{type:"json"}`),
+    { code: 'ERR_FAILED_IMPORT_ASSERTION' }
+  );
+
+  await rejects(
+    import(jsModuleDataUrl, { assert: { type: 'json' } }),
+    { code: 'ERR_FAILED_IMPORT_ASSERTION' }
+  );
+
+  {
+    const [secret0, secret1] = await Promise.all([
+      import('../fixtures/experimental.json'),
+      import(
+        '../fixtures/experimental.json',
+        { assert: { type: 'json' } }
+      ),
+    ]);
+
+    strictEqual(secret0.default.ofLife, 42);
+    strictEqual(secret1.default.ofLife, 42);
+    strictEqual(secret0.default, secret1.default);
+    // strictEqual(secret0, secret1);
```

Comment:
With the current implementation `await import('./file.json') !== await import('./file.json', {assert:{type:'json'}})`, but `(await import('./file.json')).default === (await import('./file.json', {assert:{type:'json'}})).default`, are we OK with that?

Reply:
negative, they should be the same module.

Reply:
OK, fixed in https://github.com/nodejs/node/pull/39921/commits/debc47f182e3648c736fa4162ceaa76852d2feee.